base/cvd/cuttlefish/host/commands/cvdalloc/cvdalloc.cpp

```diff
@@ -0,0 +1,145 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <string.h>
+#include <unistd.h>
+
+#include <android-base/logging.h>
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "absl/strings/str_format.h"
+
+#include "allocd/alloc_utils.h"
+#include "cuttlefish/common/libs/fs/shared_fd.h"
+
+ABSL_FLAG(int, id, 0, "Id");
+ABSL_FLAG(int, socket, 0, "Socket");
+
+namespace cuttlefish {
+namespace {
+
+void Usage() {
+  LOG(ERROR) << "cvdalloc --id=id --socket=fd ";
+  LOG(ERROR) << "Should only be invoked from run_cvd.";
+}
+
+Result<void> Allocate(int id) {
+  LOG(INFO) << "cvdalloc: allocating network resources";
+
+  CF_EXPECT(CreateBridge("cvd-br"));
+  CF_EXPECT(
+      CreateMobileIface(absl::StrFormat("cvd-mtap-%02d", id), id, kMobileIp));
+  CF_EXPECT(
+      CreateMobileIface(absl::StrFormat("cvd-wtap-%02d", id), id, kWirelessIp));
+  CF_EXPECT(CreateEthernetIface(absl::StrFormat("cvd-etap-%02d", id), "cvd-br",
+                                true, true, false));
+  return {};
+}
+
+Result<void> Teardown(int id) {
+  LOG(INFO) << "cvdalloc: tearing down resources";
+
+  DestroyMobileIface(absl::StrFormat("cvd-mtap-%02d", id), id, kMobileIp);
+  DestroyMobileIface(absl::StrFormat("cvd-wtap-%02d", id), id, kWirelessIp);
+  DestroyEthernetIface(absl::StrFormat("cvd-etap-%02d", id), true, true, false);
+  DestroyBridge("cvd-br");
+  return {};
+}
+
+}  // namespace
+
+Result<int> CvdallocMain(int argc, char *argv[]) {
+  std::vector<char *> args = absl::ParseCommandLine(argc, argv);
+
+  if (absl::GetFlag(FLAGS_id) == 0 || absl::GetFlag(FLAGS_socket) == 0) {
+    Usage();
+    /* No need to dump a trace for usage. */
+    return 1;
+  }
+
+  /*
+   * Explicit setuid calls seem to be required.
+   *
+   * This likely has something to do with invoking external commands,
+   * but it isn't clear why an explicit setuid(0) is necessary.
+   * It's possible a Linux kernel bug around permissions checking on tap
+   * devices may be the culprit, which we can't control.
+   *
+   * TODO: We should just use capabilities on Linux and rely on setuid as
+   * a fallback for other platforms.
+   *
+   * Save our current uid, so we can restore it to drop privileges later.
+   */
+  uid_t orig = getuid();
+
+  int id = absl::GetFlag(FLAGS_id);
+
+  int r = setuid(0);
+  if (r == -1) {
+    return CF_ERRNO("Couldn't setuid root: " << strerror(errno));
+  }
+
+  Result<void> result = Allocate(id);
+  if (!result.ok()) {
+    Teardown(id);
+    return CF_ERRNO("Allocation failed");
+  }
+
+  auto sock = SharedFD::Dup(absl::GetFlag(FLAGS_socket));
+  if (!sock->IsOpen()) {
+    Teardown(id);
+    return CF_ERRNO("cvdalloc: socket is closed: " << sock->StrError());
+  }
+  r = close(absl::GetFlag(FLAGS_socket));
```

Comment:
In practice this close can only fail with `EINTR` since the dup above succeeded. That could be handled by wrapping the close call like this: `TEMP_FAILURE_RETRY(close(...))`